### PR TITLE
RMTP-78 Disable Assembly Persistence for .NET 5

### PR DESCRIPTION
**Pull request by raffaelfoidl**
https://re-motion.atlassian.net/browse/RMTP-78

### CODE REVIEW

**Comment by raffaelfoidl** [comment]
ad `AssemblyGen` bullet point from ticket: there is no `AssemblyBuilder.Save()` call in this file anymore.

**Review by iam-mholle — APPROVED** [review]

File: Core/CodeGeneration/ReflectionEmit/Abstractions/AssemblyBuilderAdapter.cs [hunk 1]
```diff
@@ -49,11 +49,15 @@ public byte[] PublicKey
...
     public string SaveToDisk ()
     {
+#if FEATURE_ASSEMBLYBUILDER_SAVE
       // Scope name is the module name or file name, i.e., assembly name + '.dll'.
       _assemblyBuilder.Save (_moduleBuilder.ScopeName);
 
       // This is the absolute path to the module, which is also the assembly file path for single-module assemblies.
       return _moduleBuilder.FullyQualifiedName;
+#else
+      throw new PlatformNotSupportedException("Assembly persistence is not supported.");
```
Comment (MichaelKetting) [inline Core/CodeGeneration/ReflectionEmit/Abstractions/AssemblyBuilderAdapter.cs]:
```suggestion
      throw new PlatformNotSupportedException ("Assembly persistence is not supported.");
```

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
I'm sorry, I just realized something that wasn't spelled out in the task:

Please also update the tests for these types. The tests should not fail if Feature_AssemblyBuilder_Save is not set. You'll probably need to run the tests with the featur eflag switched off and see what needs changing.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 2]
```diff
@@ -47,6 +47,9 @@ public void SetUp ()
       _currentDirectory = Environment.CurrentDirectory;
     }
     [Test]
+#if !FEATURE_ASSEMBLYBUILDER_SAVE
+    [Ignore]
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: It think it sould be possible to update the tests by only changing the CheckAdapterAndSaveToDiskBehavior method to remove the failing parts. It might also be a good idea to just split up the Check-method into two parts with dedicated names and then not do the SaveToDiskBehavior calls in the tests. That way, one can see from the name of the called check methods what parts are tested and what are not tested.

File: Core/CodeGeneration/ReflectionEmit/ModuleBuilderFactory.cs [hunk 3]
```diff
@@ -68,7 +78,7 @@ public IModuleBuilder CreateModuleBuilder (string assemblyName, string assemblyD
     }
 
     [NotNull]
-    private StrongNameKeyPair GetKeyPair ([CanBeNull]string keyFilePathOrNull)
+    private StrongNameKeyPair GetKeyPair ([CanBeNull] string keyFilePathOrNull)
```
Comment (MichaelKetting) [inline Core/CodeGeneration/ReflectionEmit/ModuleBuilderFactory.cs]:
```suggestion
    private StrongNameKeyPair GetKeyPair ([CanBeNull]string keyFilePathOrNull)
```

unrelated change

File: IntegrationTests/Pipeline/ConcurrencyTest_WithMultipleGenerationThreads.cs [hunk 4]
```diff
@@ -125,6 +125,9 @@ public void CodeGenerationIsParallelInSeparateAssemblies_WithDegreeOfParallelism
     }
 
     [Test]
+#if !FEATURE_ASSEMBLYBUILDER_SAVE
+    [Ignore]
```
Comment (MichaelKetting) [inline IntegrationTests/Pipeline/ConcurrencyTest_WithMultipleGenerationThreads.cs]:
```suggestion
    [Ignore ("CodeManager.FlushCodeToDisk() is not supported.")]
```

File: IntegrationTests/Pipeline/FlushGeneratedCodeTest.cs [hunk 5]
```diff
@@ -28,6 +28,9 @@
 namespace Remotion.TypePipe.IntegrationTests.Pipeline
 {
   [TestFixture]
+#if !FEATURE_ASSEMBLYBUILDER_SAVE
+  [Ignore]
```
Comment (MichaelKetting) [inline IntegrationTests/Pipeline/FlushGeneratedCodeTest.cs]:
```suggestion
  [Ignore ("CodeManager.FlushCodeToDisk() is not supported.")]
```

File: IntegrationTests/Pipeline/LoadFlushedCodeTest.cs [hunk 6]
```diff
@@ -27,6 +27,9 @@
 namespace Remotion.TypePipe.IntegrationTests.Pipeline
 {
   [TestFixture]
+#if !FEATURE_ASSEMBLYBUILDER_SAVE
+  [Ignore]
```
Comment (MichaelKetting) [inline IntegrationTests/Pipeline/LoadFlushedCodeTest.cs]:
```suggestion
  [Ignore ("CodeManager.FlushCodeToDisk() is not supported.")]
```

File: IntegrationTests/Pipeline/ParticipantStateTest.cs [hunk 7]
```diff
@@ -23,6 +23,9 @@
 namespace Remotion.TypePipe.IntegrationTests.Pipeline
 {
   [TestFixture]
+#if !FEATURE_ASSEMBLYBUILDER_SAVE
+  [Ignore]
```
Comment (MichaelKetting) [inline IntegrationTests/Pipeline/ParticipantStateTest.cs]:
```suggestion
  [Ignore ("CodeManager.FlushCodeToDisk() is not supported.")]
```

File: IntegrationTests/Serialization/SerializationTestBase.cs [hunk 8]
```diff
@@ -77,6 +77,9 @@ private static IParticipant CreateCallbackImplementingParticipant ()
     }
 
     [Test]
+#if !FEATURE_ASSEMBLYBUILDER_SAVE
+    [Ignore]
```
Comment (MichaelKetting) [inline IntegrationTests/Serialization/SerializationTestBase.cs]: We should only have to disable the SimpleSerializationTest.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 9]
```diff
@@ -100,20 +100,38 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
       Assert.That (moduleBuilderAdapter.ScopeName, Is.EqualTo (c_assemblyFileName));
       Assert.That (moduleBuilderAdapter.AssemblyBuilder, Is.TypeOf<AssemblyBuilderAdapter>());
-      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
+
       Assert.That (assemblyBuilderAdapter.AssemblyName, Is.EqualTo (c_assemblyName));
       Assert.That (assemblyBuilderAdapter.PublicKey, Is.EqualTo (expectedPublicKey ?? new byte[0]));
 
-      var assemblyPath = Path.Combine (assemblyDirectory, c_assemblyFileName);
-      var pdbPath = Path.Combine (assemblyDirectory, c_pdbFileName);
       Assert.That (File.Exists (assemblyPath), Is.False);
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs:136]: These two go int othe next method.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 10]
```diff
@@ -100,20 +100,38 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
     }
 
     private void CheckAdapterAndSaveToDiskBehavior (IModuleBuilder moduleBuilder, string assemblyDirectory, byte[] expectedPublicKey = null)
+    {
+      var assemblyPath = Path.Combine (assemblyDirectory, c_assemblyFileName);
+      var pdbPath = Path.Combine (assemblyDirectory, c_pdbFileName);
+      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: The hard cast shouldn't be done before the assertion.  (line 124)

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 11]
```diff
@@ -100,20 +100,38 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
+        byte[] expectedPublicKey,
+        AssemblyBuilderAdapter assemblyBuilderAdapter,
+        string assemblyPath,
+        string pdbPath)
     {
       Assert.That (moduleBuilder, Is.TypeOf<ModuleBuilderAdapter>());
       var moduleBuilderAdapter = (ModuleBuilderAdapter) moduleBuilder;
       Assert.That (moduleBuilderAdapter.ScopeName, Is.EqualTo (c_assemblyFileName));
       Assert.That (moduleBuilderAdapter.AssemblyBuilder, Is.TypeOf<AssemblyBuilderAdapter>());
-      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs:108]: We may want to duplicate this between the two check-methods.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 12]
```diff
@@ -100,20 +100,38 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
     }
 
     private void CheckAdapterAndSaveToDiskBehavior (IModuleBuilder moduleBuilder, string assemblyDirectory, byte[] expectedPublicKey = null)
+    {
+      var assemblyPath = Path.Combine (assemblyDirectory, c_assemblyFileName);
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: Those pahts get moved to the CheckSaveToDiskBehavior  method.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 13]
```diff
@@ -100,20 +100,38 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
+      var pdbPath = Path.Combine (assemblyDirectory, c_pdbFileName);
+      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
+
+      CheckAdapterBehavior (moduleBuilder, expectedPublicKey, assemblyBuilderAdapter, assemblyPath, pdbPath);
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: The thought behind the rewrite was that there's only two method calls left and the CheckAdapterAndSaveToDiskBehavior  can be inlined.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 14]
```diff
@@ -96,23 +110,35 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
+      Assert.That (moduleBuilder, Is.TypeOf<ModuleBuilderAdapter>());
+      var moduleBuilderAdapter = (ModuleBuilderAdapter) moduleBuilder;
+      Assert.That (moduleBuilderAdapter.ScopeName, Is.EqualTo (c_assemblyFileName));
+      Assert.That (moduleBuilderAdapter.AssemblyBuilder, Is.TypeOf<AssemblyBuilderAdapter>());
+      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
 
       var assemblyPath = Path.Combine (assemblyDirectory, c_assemblyFileName);
       var pdbPath = Path.Combine (assemblyDirectory, c_pdbFileName);
-      Assert.That (File.Exists (assemblyPath), Is.False);
-      Assert.That (File.Exists (pdbPath), Is.False);
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: Please re-add those two lines. Removing them would be an unrelated change and you'd have to consider if the event of a left-over file could result in a false positve or strange exception.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 15]
```diff
@@ -96,23 +110,35 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
+
       Assert.That (assemblyBuilderAdapter.AssemblyName, Is.EqualTo (c_assemblyName));
       Assert.That (assemblyBuilderAdapter.PublicKey, Is.EqualTo (expectedPublicKey ?? new byte[0]));
+    }
+
+    private void CheckSaveToDiskBehavior (IModuleBuilder moduleBuilder, string assemblyDirectory)
+    {
+      Assert.That (moduleBuilder, Is.TypeOf<ModuleBuilderAdapter>());
+      var moduleBuilderAdapter = (ModuleBuilderAdapter) moduleBuilder;
+      Assert.That (moduleBuilderAdapter.ScopeName, Is.EqualTo (c_assemblyFileName));
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]:
```suggestion
```

This line is already covered by the other test.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 16]
```diff
@@ -96,23 +110,35 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
+      Assert.That (moduleBuilderAdapter.AssemblyBuilder, Is.TypeOf<AssemblyBuilderAdapter>());
+      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
 
       var assemblyPath = Path.Combine (assemblyDirectory, c_assemblyFileName);
       var pdbPath = Path.Combine (assemblyDirectory, c_pdbFileName);
-      Assert.That (File.Exists (assemblyPath), Is.False);
-      Assert.That (File.Exists (pdbPath), Is.False);
 
       var result = assemblyBuilderAdapter.SaveToDisk();
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]:
```suggestion
      var result = moduleBuilder.AssemblyBuilder.SaveToDisk();
```

Given the new test-layout, we no longer need to do a cast to AssemblyBuilderAdapter but isntead we can just use SaveToDisk from the interface. I normally wouldn't do this change, but it allows us to remove the redundant assertions in line 137/138.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 17]
```diff
@@ -96,23 +110,35 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
       Assert.That (assemblyBuilderAdapter.PublicKey, Is.EqualTo (expectedPublicKey ?? new byte[0]));
+    }
+
+    private void CheckSaveToDiskBehavior (IModuleBuilder moduleBuilder, string assemblyDirectory)
+    {
+      Assert.That (moduleBuilder, Is.TypeOf<ModuleBuilderAdapter>());
+      var moduleBuilderAdapter = (ModuleBuilderAdapter) moduleBuilder;
+      Assert.That (moduleBuilderAdapter.ScopeName, Is.EqualTo (c_assemblyFileName));
+      Assert.That (moduleBuilderAdapter.AssemblyBuilder, Is.TypeOf<AssemblyBuilderAdapter>());
+      var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]:
```suggestion
```
Not needed here. See comment with SafeToDisk().

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 18]
```diff
@@ -96,23 +110,35 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
       var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
+
       Assert.That (assemblyBuilderAdapter.AssemblyName, Is.EqualTo (c_assemblyName));
       Assert.That (assemblyBuilderAdapter.PublicKey, Is.EqualTo (expectedPublicKey ?? new byte[0]));
+    }
+
+    private void CheckSaveToDiskBehavior (IModuleBuilder moduleBuilder, string assemblyDirectory)
+    {
+      Assert.That (moduleBuilder, Is.TypeOf<ModuleBuilderAdapter>());
+      var moduleBuilderAdapter = (ModuleBuilderAdapter) moduleBuilder;
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]:
```suggestion
```
Not needed here. See comment with SafeToDisk().

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
Was able to remove some redundant code.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs (hunk 14 above)
Comment (raffaelfoidl) [reply Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: I'm afraid I don't quite understand. In the last review you suggested moving the declaration of `assemblyPath` and `pdbPath` to `CheckSaveToDiskBehavior`. In the course of doing so, these two assertions also moved into this method (lines 140/141 in the most current commit). Or am I misunderstanding something?

Comment (raffaelfoidl) [reply Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: I have to admit that I had to actively search for these two lines in the GitHub diff view by expanding some collapsed lines - but I knew they were there, otherwise I would not have done so probably.

Comment (raffaelfoidl) [reply Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: Oh no. I simply can't read and distinguish `true` from `false`. sorry

Comment (raffaelfoidl) [reply Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]: don't know how those poor lines got lost :( but now they are back and the spam is over.

File: Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs [hunk 19]
```diff
@@ -96,25 +110,34 @@ public void CreateModuleBuilder_StrongNamed_ProvidedKey ()
...
       Assert.That (moduleBuilderAdapter.ScopeName, Is.EqualTo (c_assemblyFileName));
       Assert.That (moduleBuilderAdapter.AssemblyBuilder, Is.TypeOf<AssemblyBuilderAdapter>());
       var assemblyBuilderAdapter = (AssemblyBuilderAdapter) moduleBuilder.AssemblyBuilder;
+
       Assert.That (assemblyBuilderAdapter.AssemblyName, Is.EqualTo (c_assemblyName));
       Assert.That (assemblyBuilderAdapter.PublicKey, Is.EqualTo (expectedPublicKey ?? new byte[0]));
+    }
+
+    private void CheckSaveToDiskBehavior (IModuleBuilder moduleBuilder, string assemblyDirectory)
+    {
```
Comment (MichaelKetting) [inline Core.UnitTests/CodeGeneration/ReflectionEmit/ModuleBuilderFactoryTest.cs]:
```suggestion
```

sorry, missed the blank line

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Review by MichaelKetting — APPROVED** [review]
